includes/Core/DI/DI_Container.php

```diff
@@ -0,0 +1,196 @@
+<?php
+/**
+ * Class Google\Site_Kit\Core\DI\DI_Container
+ *
+ * @package   Google\Site_Kit\Core\DI
+ * @copyright 2021 Google LLC
+ * @license   https://www.apache.org/licenses/LICENSE-2.0 Apache License 2.0
+ * @link      https://sitekit.withgoogle.com
+ */
+
+namespace Google\Site_Kit\Core\DI;
+
+use ArrayAccess;
+use Google\Site_Kit_Dependencies\Psr\Container\ContainerInterface;
+
+/**
+ * DI container class.
+ *
+ * @since n.e.x.t
+ */
+class DI_Container implements ContainerInterface, ArrayAccess {
+
+	/**
+	 * Definitions list.
+	 *
+	 * @since n.e.x.t
+	 * @var array
+	 */
+	protected $definitions = array();
+
+	/**
+	 * Returns true if the container can return an entry for the given identifier.
+	 * Returns false otherwise.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $id Identifier of the entry to look for.
+	 * @return bool TRUE if the container can return an entry, otherwise FALSE.
+	 */
+	public function offsetExists( $id ) {
+		return $this->has( $id );
+	}
+
+	/**
+	 * Finds an entry of the container by its identifier and returns it.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $id Identifier of the entry to look for.
+	 * @return mixed Entry.
+	 */
+	public function offsetGet( $id ) {
+		return $this->get( $id );
+	}
+
+	/**
+	 * Sets the entry definition.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $id Identifier of the entry.
+	 * @param mixed  $entry Entry.
+	 */
+	public function offsetSet( $id, $entry ) {
+		$this->set( $id, $entry );
+	}
+
+	/**
+	 * Removes the entry.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $id Identifier of the entry to remove.
+	 */
+	public function offsetUnset( $id ) {
+		unset( $this->definitions[ $id ] );
+	}
+
+	/**
+	 * Sets the entry definition.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $id Identifier of the entry.
+	 * @param mixed  $entry Entry.
+	 * @return bool TRUE if the entry is added, otherwise FALSE.
+	 */
+	public function set( $id, $entry ) {
+		if ( ! empty( $this->definitions[ $id ]['is_protected'] ) ) {
+			return false;
+		}
+
+		$this->definitions[ $id ] = array(
+			'is_factory'   => false,
+			'is_protected' => false,
+			'is_service'   => is_callable( $entry ),
+			'entry'        => $entry,
+		);
+
+		return true;
+	}
+
+	/**
+	 * Sets the entry to be protected.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $id Identifier of the entry.
+	 */
+	public function set_protected( $id ) {
+		if ( $this->has( $id ) ) {
+			$this->definitions[ $id ]['is_protected'] = true;
+		}
+	}
+
+	/**
+	 * Sets the entry to be factory.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $id Identifier of the entry.
+	 */
+	public function set_factory( $id ) {
```

Comment:
This may not be needed if we go with the approach of my first comment. IMO there are three types of entries:
* An entry that just returns whatever value is set. This is the default. Can be set using `set_value`.
* A service entry, which returns the return value from a callback, invoked just once. Can be set using `set_service`.
* A factory entry, which returns the return value from a callback, on every access call. Can be set using `set_factory`.

Any of those three types can be protected, so keeping a `set_protected` (or `set_is_protected`, see above) around makes sense.

Reply:
Yep, updated.